pkg/config/signer.go

```diff
@@ -0,0 +1,225 @@
+		cert:       cert,
+		mspID:      mspID,
+		privateKey: privateKey,
+		publicKey:  publicKey,
+	}
+	return signer, nil
+}
+
+// Serialize returns a byte array representation of this identity.
+func (s *Signer) Serialize() ([]byte, error) {
+	pb := &pem.Block{Bytes: s.cert.Raw, Type: "CERTIFICATE"}
+	pemBytes := pem.EncodeToMemory(pb)
+	if pemBytes == nil {
+		return nil, errors.New("failed to encode pem block")
+	}
+
+	// serialize identities by prepending the MSPID and appending
+	// the ASN.1 DER content of the cert
+	sID := &msp.SerializedIdentity{Mspid: s.mspID, IdBytes: pemBytes}
+	idBytes := MarshalOrPanic(sID)
+
+	return idBytes, nil
+}
+
+// Public returns the public key of the signer.
+func (s *Signer) Public() crypto.PublicKey {
+	return s.publicKey
+}
+
+// Cert returns the certificate of the signer.
+func (s *Signer) Cert() *x509.Certificate {
+	return s.cert
+}
+
+// MSPId returns the MSP ID of the signer.
+func (s *Signer) MSPId() string {
+	return s.mspID
+}
+
+// Sign performs ECDSA sign with signer's private key on given digest. It
+// ensures signatures are created with Low S values since Fabric normalizes
+// all signatures to Low S.
+// See https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki#low_s
+// for more detail.
+func (s *Signer) Sign(reader io.Reader, digest []byte) (signature []byte, err error) {
+	if reader == nil {
+		return nil, errors.New("failed to sign, reader can not be nil")
+	}
+
+	rr, ss, err := ecdsa.Sign(reader, s.privateKey, digest)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure Low S signatures
+	sig := toLowS(
+		s.privateKey.PublicKey,
+		ecdsaSignature{
+			R: rr,
+			S: ss,
+		},
+	)
+
+	return asn1.Marshal(sig)
+}
+
+// CreateSignatureHeader returns a new SignatureHeader. It contains a valid
+// nonce and a creator, which is a serialized representation of the signer.
+func (s *Signer) CreateSignatureHeader() (*common.SignatureHeader, error) {
+	creator, err := s.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	nonce, err := createNonce()
+	if err != nil {
+		return nil, err
+	}
+
+	return &common.SignatureHeader{
+		Creator: creator,
+		Nonce:   nonce,
+	}, nil
+}
+
+// createNonce generates a nonce using the crypto/rand package.
+func createNonce() ([]byte, error) {
+	nonce, err := getRandomNonce()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random nonce: %s", err)
+	}
+	return nonce, nil
+}
+
+func getRandomNonce() ([]byte, error) {
+	key := make([]byte, 24)
+
+	_, err := rand.Read(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get random bytes: %s", err)
+	}
+	return key, nil
+}
+
+// When using ECDSA, both (r,s) and (r, -s mod n) are valid signatures.
+// In order to protect against signature malleability attacks, Fabric
+// normalizes all signatures to a canonical form where s is at most half
+// the order of the curve. In order to make signatures compliant with what
+// Fabric expects, toLowS creates signatures in this canonical form.
```

Comment:
This makes me wonder - why are we enforcing this on signatures produced by leaf certificates like peers or orderers?
I understand why we need this for CA signatures on identities, but - what is the attack we're protecting against here? Duplication of SignedData (triplet of message, signature and identity)? Seems unlikely since I can anyway sign as many times as I want.

Reply:
I also don't see any issues with endorser signatures; at the same time, it's more uniform to have the same approach everywhere, and we don't know what new type of signature leaf nodes might be required to produce in the future, so it seems prudent to me to stick with this restriction.